### PR TITLE
Piggyback on the first traversal and capture links to parents in Workspace Context !

### PR DESCRIPTION
**Pull request by TilakMaddy**
We piggyback on the first traversal. So imagine that you are a node. 

Now you "tell" the ASTConstVisitors 2 things when "accept" is called on you - 
a) your id
b) ids of all your children

Then the ASTConstVisitor can maintain a mapping from  b -> a which will be the `parent_link` 

So that's exactly what happens in this PR.  Link to important snippets and explanation below

### CODE REVIEW

File: aderyn_core/src/context/workspace_context.rs [hunk 1]
```diff
@@ -1962,6 +1967,17 @@ impl ASTConstVisitor for WorkspaceContext {
         );
         Ok(true)
     }
+
+    fn visit_immediate_children(
```
Comment (TilakMaddy) [inline aderyn_core/src/context/workspace_context.rs:1971]: As the ASTConstVisitor you maintain b -> a mapping as described in PR readme

File: aderyn_core/src/context/workspace_context.rs [hunk 2]
```diff
@@ -742,6 +742,7 @@ pub struct WorkspaceContext {
     last_contract_definition_id: Option<NodeID>,
     last_function_definition_id: Option<NodeID>,
     last_modifier_definition_id: Option<NodeID>,
+    pub parent_link: HashMap<NodeID, NodeID>,
```
Comment (TilakMaddy) [inline aderyn_core/src/context/workspace_context.rs:745]: Here are the links populated in the first traversal itself

File: aderyn_core/src/detect/experimental/parent_chain.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+#[derive(Default)]
+pub struct ParentChainDemonstrator {
+    // Keys are source file name and line number
+    found_instances: BTreeMap<(String, usize), NodeID>,
+}
+
+/*
+
+In ParentChainContract.sol, there is only 1 assignment done. The goal is to capture it first, second and third parent
```
Comment (TilakMaddy) [inline aderyn_core/src/detect/experimental/parent_chain.rs:20]: This whole thing is the example

File: tests/contract-playground/src/parent_chain/ParentChainContract.sol [hunk 4]
```diff
@@ -0,0 +1,48 @@
+// SPDX-License-Identifier: MIT
```
Comment (TilakMaddy) [inline tests/contract-playground/src/parent_chain/ParentChainContract.sol:1]: This is a newly introduced contract to support the example

**Comment by TilakMaddy** [comment]
Improvement in next PR can be:

`context.get_parent(node.id)` -> `node.parent(context)` 

I could take it up if this is approved!

**Comment by alexroan** [comment]
@TilakMaddy, this is excellent!

A couple of follow-ups:
* `context.get_parent(node.id)` -> `node.parent(context)` (you take this one)
* Remove all the references and dependencies to the `GetParent` pattern (I'll take this one)

**Review by alexroan — APPROVED** [review]
Excellent stuff Tilak!!

**Comment by TilakMaddy** [comment]
> Excellent stuff Tilak!!

You're always welcome Alex !

**Pull request merged**